### PR TITLE
Perf: Add virtual Notebook for delayed cell rendering

### PR DESCRIPTION
**Pull request by goanpeca**
Port performance changes from 2.x branch to master (3.x)

## References

Includes the following work:

- https://github.com/jupyterlab/jupyterlab/pull/8972/files
- https://github.com/jupyterlab/jupyterlab/pull/9149/files
- https://github.com/jupyterlab/jupyterlab/pull/10001/files

### CODE REVIEW

**Comment by goanpeca** [comment]
@afshin could you make the test notebook to check on this :) ?

Thanks!

**Review by echarles — APPROVED** [review]
Reviewed and tested. LGTM. @afshin Anything else we miss to get this merged?

**Comment by echarles** [comment]
@afshin I have just discussed with @goanpeca on this PR and it looks that you may have a notebook that you'd like to share so we can test and see what it gives in terms of performance.

Do you think you can provide it soon? I guess it may contain private information and this is not often easy to remove. If this give issues, maybe you can share the pattern you'd like us to investigate (e.g. 1000 cells with some matplotlib outputs... or something else ?).

Thx, Eric

**Review by afshin — APPROVED** [review]
I was unable to generate a publicly shareable version of a notebook I've seen that was slow in 2.3.0, but in personal testing, I can see its performance is improved. It's a 4MB notebook file with very large text outputs.

I think we should merge this and then iterate on addressing a couple performance issues:

1. Switching tabs in JupyterLab with one or more large notebooks open is very slow.
2. If you grab the scroll handle in a very large notebook with virtual rendering and you drag the scroll handle in order to scroll the notebook, the notebook panel flashes a bright color. We should modify this to something more subdued.

These don't need to block the current PR and can be addressed in subsequent PRs.

**Comment by echarles** [comment]
Thx you very much @afshin for taking time to review and investigate this.

> Switching tabs in JupyterLab with one or more large notebooks open is very low.

Absolutely, this is a issue we have tried to solve with what we call a  `window`variant of the virtual notebook, but felt into issue to keep the state of the output widgets. We can relaunch this if we don't find a better way.

> If you grab the scroll handle in a very large notebook with virtual rendering and you drag the scroll handle in order to scroll the notebook, the notebook panel flashes a bright color. We should modify this to someone more subdued.

This may be the case when the cells are still being rendered and you scroll down fast beyond them. We can fine-tune in subsequent PRS.

Globally, I am sure the users of heavy notebooks will get substantial gains from this PR. Thx again to @goanpeca and all others working on this.

**Comment by mlucool** [comment]
@afshin 
>  It's a 4MB notebook file with very large text outputs.

In theory https://github.com/jupyterlab/jupyterlab/pull/9594 fixes this. But we found that https://github.com/jupyterlab/jupyterlab/issues/10046 remains a big difference vs. what limit output did in notebook.

If you have 4MB notebooks, you should be sure to install `fastjsonschema` to save on the time it takes for the server to valdiate the notebook (https://github.com/jupyter/nbformat#using-a-different-json-schema-validator).

**Comment by afshin** [comment]
Thanks for the suggestions, Marc 🚀

File: packages/cells/src/widget.ts [hunk 1]
```diff
@@ -700,32 +707,33 @@ export class CodeCell extends Cell<ICodeCellModel> {
...
+    if (!options.placeholder) {
+      // Insert the output before the cell footer.
+      const outputWrapper = (this._outputWrapper = new Panel());
+      outputWrapper.addClass(CELL_OUTPUT_WRAPPER_CLASS);
+      const outputCollapser = new OutputCollapser();
+      outputCollapser.addClass(CELL_OUTPUT_COLLAPSER_CLASS);
+      const output = (this._output = new OutputArea({
+        model: model.outputs,
+        rendermime,
+        contentFactory: contentFactory
```
Comment (krassowski) [inline packages/cells/src/widget.ts:719]: I think that we are missing `maxNumberOutputs` here. It seems it was left over when resolving conflicts https://github.com/goanpeca/jupyterlab/commit/ca2f6d886419224a1af1a6d2bc5e258593949256.

**Comment by goanpeca** [comment]
>I think that we are missing maxNumberOutputs here. It seems it was left over when resolving conflicts goanpeca@ca2f6d8.

Will fix @krassowski !

**Pull request merged**